### PR TITLE
Migrate domainName field from text type to links type

### LINKED ISSUE #5759: Migrate Domain Field to the new Link Field Type
# Current Behavior

The domain field on the company object uses a "Text" field type.

CleanShot 2024-06-06 at 10 26 52

# Desired Behavior

We want to migrate the domain field to the newly introduced linked field type, allowing users to specify multiple values (several domains).

CleanShot 2024-06-06 at 10 27 14

We should adapt the email syncing logic to account for companies having different domains attached. We should sync emails from any matching domain name (e.g., tim@apple.com or tim@apple.dev).

We should keep the custom logic for the domain field, such as fetching the favicon.

### PR DESCRIPTION
**Pull request by ijreilly**
Closes #5759.

### CODE REVIEW

File: packages/twenty-server/src/utils/getCompanyDomainName.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+// temporary, to remove once domainName has been fully migrated to Links type
+export const getCompanyDomainName = (company: any) => {
```
Comment (Weiko) [inline packages/twenty-server/src/utils/getCompanyDomainName.ts:2]: Seems this one is not aligned with getCompanyDomainName from front package

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/interfaces/field-metadata-settings.interface.ts [hunk 2]
```diff
@@ -10,6 +10,10 @@ type FieldMetadataDefaultSettings = {
   isForeignKey?: boolean;
 };
 
+type FieldMetadataLinksSettings = {
+  isDomain?: boolean;
```
Comment (Weiko) [inline packages/twenty-server/src/engine/metadata-modules/field-metadata/interfaces/field-metadata-settings.interface.ts:14]: Do we want to introduce this in this PR? I don't see where this is used

File: packages/twenty-server/src/engine/core-modules/quick-actions/quick-actions.service.ts [hunk 3]
```diff
@@ -159,7 +161,7 @@ export class QuickActionsService {
     }
 
     const enrichedData = await this.intelligenceService.enrichCompany(
-      company.domainName,
+      getCompanyNameFromDomainName(getCompanyDomainName(company.domainName)),
```
Comment (Weiko) [inline packages/twenty-server/src/engine/core-modules/quick-actions/quick-actions.service.ts:164]: It used to work with domainName only, why do we want the name of the company now?

File: packages/twenty-server/src/database/commands/upgrade-version/0-23/0-23-migrate-domain-to-links.command.ts [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    workspaceQueryRunner,
+    dataSourceMetadata,
+  }: {
+    sourceColumnName: string;
+    targetColumnName: string;
+    workspaceQueryRunner: QueryRunner;
+    dataSourceMetadata: DataSourceEntity;
+  }) {
+    await workspaceQueryRunner.query(
+      `UPDATE "${dataSourceMetadata.schema}"."company" SET "${targetColumnName}" = CASE WHEN "${sourceColumnName}" LIKE 'https://%' THEN "${sourceColumnName}" ELSE 'https://' || "${sourceColumnName}" END;`,
```
[the remark this answers is not in the document]
Comment (ijreilly) [reply packages/twenty-server/src/database/commands/upgrade-version/0-23/0-23-migrate-domain-to-links.command.ts:305]: Yes I think we really want to have that https:// prefix, we add it in the front-end.
Maybe I could replace `LIKE 'https://%'` with `LIKE 'http'` to be extra safe

File: packages/twenty-server/src/modules/company/repositories/company.repository.ts [hunk 5]
```diff
@@ -67,11 +69,11 @@ export class CompanyRepository {
     );
 
     await this.workspaceDataSourceService.executeRawQuery(
-      `INSERT INTO ${dataSourceSchema}.company (id, "domainName", name, "addressAddressCity", position)
+      `INSERT INTO ${dataSourceSchema}.company (id, "${companyDomainNameColumnName}", name, "addressAddressCity", position)
       VALUES ($1, $2, $3, $4, $5)`,
       [
         companyToCreate.id,
-        companyToCreate.domainName,
+        'https://' + companyToCreate.domainName,
```
Comment (bosiraphael) [inline packages/twenty-server/src/modules/company/repositories/company.repository.ts:76]: Why do we add 'https://' here but we don't in `getExistingCompaniesByDomainNames`? The get and the create should have a consistant api, either `companyToCreate` and `domainNames` should both contain `https://` or they shouldn't.

File: packages/twenty-server/src/engine/workspace-manager/standard-objects-prefill-data/company.ts [hunk 6]
```diff
@@ -23,7 +23,7 @@ export const companyPrefillData = async (
     .values([
       {
         name: 'Airbnb',
-        domainName: 'airbnb.com',
+        domainNamePrimaryLinkUrl: 'airbnb.com',
```
Comment (bosiraphael) [inline packages/twenty-server/src/engine/workspace-manager/standard-objects-prefill-data/company.ts:26]: `https://` ?

File: packages/twenty-server/src/database/commands/upgrade-version/0-23/0-23-migrate-domain-to-links.command.ts [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+          // Migrate data from domainName to primaryLinkUrl
+          await this.migrateDataWithinCompanyTable({
+            sourceColumnName: `${domainNameField.name}`,
+            targetColumnName: `${tmpNewDomainLinksField.name}PrimaryLinkUrl`,
+            workspaceQueryRunner,
+            dataSourceMetadata,
+          });
+
+          // Backfill primaryLinkLabel with empty string
+          await this.backfillColumnWithEmptyString({
```
[the remark this answers is not in the document]
Comment (ijreilly) [reply packages/twenty-server/src/database/commands/upgrade-version/0-23/0-23-migrate-domain-to-links.command.ts:157]: Oh you are right ! I added the defaultValue logic after, so until then I needed this. removing it now thanks

File: packages/twenty-server/src/engine/metadata-modules/field-metadata/interfaces/field-metadata-settings.interface.ts (hunk 2 above)
Comment (ijreilly) [reply packages/twenty-server/src/engine/metadata-modules/field-metadata/interfaces/field-metadata-settings.interface.ts:14]: Right it's not actually

File: packages/twenty-server/src/engine/core-modules/quick-actions/quick-actions.service.ts (hunk 3 above)
Comment (ijreilly) [reply packages/twenty-server/src/engine/core-modules/quick-actions/quick-actions.service.ts:164]: its a mistake, thanks for spotting this. (i think this service is not in use anymore by the way)

File: packages/twenty-server/src/utils/getCompanyDomainName.ts (hunk 1 above)
Comment (ijreilly) [reply packages/twenty-server/src/utils/getCompanyDomainName.ts:2]: right, thanks

File: packages/twenty-server/src/modules/company/repositories/company.repository.ts (hunk 5 above)
Comment (ijreilly) [reply packages/twenty-server/src/modules/company/repositories/company.repository.ts:76]: I had not paid attention to that, thanks. 
I am replacing the query to compare the domain names without the  'https://' or 'http://' prefix; this seems safer than adding the prefix the domainNames and comparing. (although when we migrate the existing data we are adding https:// so all existing data should have a https:// prefix but still)

Comment (ijreilly) [reply packages/twenty-server/src/modules/company/repositories/company.repository.ts:76]: yet to test though, waiting on your help on how to trigger sync again

Comment (ijreilly) [reply packages/twenty-server/src/modules/company/repositories/company.repository.ts:76]: tested !

**Review by Weiko — APPROVED** [review]
LGTM, great job @ijreilly! 👏

**Review by bosiraphael — APPROVED** [review]
LGTM

**Pull request merged**